lsp-mode.el

```diff
@@ -4511,16 +4580,34 @@ Added to `after-change-functions'."
              lsp-managed-mode)
     (run-hooks 'lsp-on-change-hook)))
 
-(defun lsp--after-change (&rest _)
+(defun lsp--on-change-fns-debounce ()
+  (when (and (not lsp-inhibit-lsp-hooks)
+             lsp-managed-mode)
+    (run-hook-with-args 'lsp-on-change-functions lsp--changes)
+    (setq lsp--changes nil)))
+
+(defun lsp--after-change (buffer start end pre-change-len)
   (when lsp--on-change-timer
     (cancel-timer lsp--on-change-timer))
-
   (setq lsp--on-change-timer (run-with-idle-timer
                               lsp-idle-delay
                               nil
                               #'lsp--on-change-debounce
-                              (current-buffer)))
-  (lsp--idle-reschedule (current-buffer)))
+                              buffer))
+
+  (when lsp--on-change-fns-timer
+    (cancel-timer lsp--on-change-fns-timer)
+    (setq lsp--changes
+          (plist-put lsp--changes
+                     buffer
+                     (nconc (plist-get lsp--changes buffer)
+                            `((,start ,end ,pre-change-len))))))
```

Comment:
don't we need to keep only the first start?

Reply:
Yes, but since I add new `lsp-on-change-functions`, it might be better to provide all information to its subscribers.
Thus all information will be passed.

Reply:
> it might be better to provide all information to its subscribers.

Why? Can we implement the simplest possible solution and extend it when a subscriber that needs more information is needed? More code = more trouble in the future.

Reply:
Also, this will break if there are 2 calls that need the first edited point. The min position must be stored in the capture, not in buffer-local.

Reply:
It actually less code, we just store whatever `after-change-functions` gave us, without further modification.
There's no min position here yet, it's just a series of changes and the subscribers need to process more to get information they needed.
The main reason I'm not processing more than this is I don't know what exactly subscriber want (generally), they may want  only first edit point, or a range, put just raw data here provide the max flexibility.

Reply:
We have one subscriber and we know what it needs. Either way, it should not interfere with the rest of the "subscribers" since they will have a scope. 

``` emacs-lisp
(let ((first-edited (point-max)))
  (add-hook 'before-change-functions 
            (lambda (start &rest _)
              (setq first-edited (min start first-edited)))
            nil t)

  (lsp-request-async 
   ...
   (lambda (edits)
     ;; check if edits is compatible with first-edited point
     )))
```

Reply:
Well, I've added a new hook `lsp-on-change-functions` and want to provide maximum flexibility for its potential subscribers.
It looks like that's an over engineering attempt, will revert that.

Reply:
@yyoncho